packages/Teams/Middleware/SSOTokenExchangeMiddleware.cs

```diff
@@ -0,0 +1,210 @@
+    /// deduplication.
+    /// </remarks>
+    public class SSOTokenExchangeMiddleware : IMiddleware
+    {
+        private const string ConnectionNameProperty = "ConnectionName";
+        private readonly IStorage _storage;
+        private readonly string _oAuthConnectionName;
+
+        /// <summary>
+        /// Creates an instance of <see cref="SSOTokenExchangeMiddleware" />.
+        /// </summary>
+        /// <param name="configuration"></param>
+        /// <param name="storage"></param>
+        public SSOTokenExchangeMiddleware(IConfiguration configuration, IStorage storage)
+        {
+            _oAuthConnectionName = configuration[ConnectionNameProperty];
+            _storage = storage;
+        }
+
+        public async Task OnTurnAsync(ITurnContext turnContext, NextDelegate next, CancellationToken cancellationToken = default)
+        {
+            if (turnContext.Activity.Name == SignInConstants.TokenExchangeOperationName)
+            {
+                // If the TokenExchange is NOT successful, the response will have already been sent by ExchangedTokenAsync
+                if (!await this.ExchangedTokenAsync(turnContext, cancellationToken).ConfigureAwait(false))
+                {
+                    return;
+                }
+
+                // Only one token exchange should proceed from here. Deduplication is performed second because in the case
+                // of failure due to consent required, every caller needs to receive the 
+                if (!await DeduplicatedTokenExchangeIdAsync(turnContext, cancellationToken))
+                {
+                    // If the token is not exchangeable, do not process this activity further.
+                    return;
+                }
+            }
+
+            await next(cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task<bool> DeduplicatedTokenExchangeIdAsync(ITurnContext turnContext, CancellationToken cancellationToken)
+        {
+            // Create a StoreItem with Etag of the unique 'signin/tokenExchange' request
+            var storeItem = new TokenStoreItem
+            {
+                ETag = (turnContext.Activity.Value as JObject).Value<string>("id")
+            };
+
+            var storeItems = new Dictionary<string, object> { { TokenStoreItem.GetStorageKey(turnContext), storeItem } };
+            try
+            {
+                // Writing the IStoreItem with ETag of unique id will succeed only once
+                await _storage.WriteAsync(storeItems, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception ex)
+
+                // Memory storage throws a generic exception with a Message of 'Etag conflict. [other error info]'
+                // CosmosDbPartitionedStorage throws: ex.Message.Contains("pre-condition is not met")
+                when (ex.Message.StartsWith("Etag conflict") || ex.Message.Contains("pre-condition is not met"))
```

Comment:
This seems a little unreliable. Can we not check for a particular exception type and some kind of error code instead?

Reply:
I don't like this either.